codaledgercli/schnorr.py

```diff
@@ -0,0 +1,247 @@
+import hashlib
+import binascii
+import sys
+from random import getrandbits
+from codaledgercli import poseidon
+
+a = 0
+b = 7
+a_coeff = a
+b_coeff = b
+
+p = 5543634365110765627805495722742127385843376434033820803590214255538854698464778703795540858859767700241957783601153
+n = 5543634365110765627805495722742127385843376434033820803592568747918351978899288491582778380528407187068941959692289
+G = [1, 1587713460471950740217388326193312024737041813752165827005856534245539019723616944862168333942330219466268138558982]
+N = 382
+
+def next_bit():
+    return getrandbits(1)
+
+def legendre_symbol(a):
+    ls = pow(a, (p - 1)//2, p)
+    if ls == p - 1:
+        return -1
+    return ls
+
+def prime_mod_sqrt(a):
+    """
+    Square root modulo prime number
+    Solve the equation
+        x^2 = a mod p
+    and return list of x solution
+    http://en.wikipedia.org/wiki/Tonelli-Shanks_algorithm
+    """
+    a %= p
+
+    # Simple case
+    if a == 0:
+        return [0]
+    if p == 2:
+        return [a]
+
+    # Check solution existence on odd prime
+    if legendre_symbol(a) != 1:
+        return []
+
+    # Simple case
+    if p % 4 == 3:
+        x = pow(a, (p + 1)//4, p)
+        return [x, p-x]
+
+    # Factor p-1 on the form q * 2^s (with Q odd)
+    q, s = p - 1, 0
+    while q % 2 == 0:
+        s += 1
+        q //= 2
+
+    # Select a z which is a quadratic non resudue modulo p
+    z = 1
+    while legendre_symbol(z) != -1:
+        z += 1
+    c = pow(z, q, p)
+
+    # Search for a solution
+    x = pow(a, (q + 1)//2, p)
+    t = pow(a, q, p)
+    m = s
+    while t != 1:
+        # Find the lowest i such that t^(2^i) = 1
+        i, e = 0, 2
+        for i in range(1, m):
+            if pow(t, e, p) == 1:
+                break
+            e *= 2
+
+        # Update next value to iterate
+        b = pow(c, 2**(m - i - 1), p)
+        x = (x * b) % p
+        t = (t * b * b) % p
+        c = (b * b) % p
+        m = i
+
+    return [x, p-x]
+
+def random_field_elt():
+    res = 0
+    for i in range(N):
+        res += next_bit() * (2 ** i)
```

Comment:
these files are only needed for my testing the new curves because we don't have an ocaml implementation of them to test against yet -- i moved them into /test, sorry that this wasn't explicit. only __main__.py and decode.py need to be distributed to and called by users